plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -140,26 +141,26 @@ public CacheStatsMBean getMetadataCacheStats()
     }
 
     public TableSnapshot loadSnapshot(SchemaTableName table, Path tableLocation, ConnectorSession session)
-            throws IOException
+            throws IOException, ExecutionException
     {
         String location = tableLocation.toString();
         TableSnapshot cachedSnapshot = tableSnapshots.getIfPresent(location);
         TableSnapshot snapshot;
         FileSystem fileSystem = getFileSystem(tableLocation, table, session);
         if (cachedSnapshot == null) {
-            snapshot = TableSnapshot.load(
-                    table,
-                    fileSystem,
-                    tableLocation,
-                    parquetReaderOptions,
-                    checkpointRowStatisticsWritingEnabled);
-            tableSnapshots.put(location, snapshot);
+            snapshot = tableSnapshots.get(location, () ->
```

Comment:
Until we get to this point the snapshot could already be existing in the `tableSnapshots` cache.


A possible alternative would be to have a way to safely know whether the snapshot is already cached and act accordingly.

```
        AtomicBoolean isTableSnapshotAlreadyCached = new AtomicBoolean(false);
        TableSnapshot cachedTableSnapshot = tableSnapshots.get(location, () ->
        {
            isTableSnapshotAlreadyCached.set(false);
            return TableSnapshot.load(
                    table,
                    fileSystem,
                    tableLocation,
                    parquetReaderOptions,
                    checkpointRowStatisticsWritingEnabled);
        });
```

Reply:
> Until we get to this point the snapshot could already be existing in the tableSnapshots cache.

And if this is the case we will get it from there instead of loading it.

Reply:
>  AtomicBoolean isTableSnapshotAlreadyCached = new AtomicBoolean(false);

I assume you made a mistake and initial value would be true, right ? 
And the point of this part is to change the implementation in such a way we call `cache.get` only once? Do I understand correctly?

Reply:
> I assume you made a mistake and initial value would be true, right ?
Yes, I've corrected the snipped.

> we call `cache.get` only once
Yes, that's my point.

Reply:
I can do that but I am not sure that will be cleaner than the current one. I find them both similarly ugly to follow ;) 
@findepi  WDYT ?

Reply:
will take another look once https://github.com/trinodb/trino/pull/11562#pullrequestreview-915838127 gets applied